File: Week01/d3_permutation.py
"""
Author: Penut Chen
Date: 2019-6-29
Description:
  Practice of permutation function.
"""

def main():
    size = 4
    arr = list([a + 1 for a in range(size)])
    count = [0]
    permutation(arr, 0, count)
    print('Total %d' % count[0])

def permutation(arr, idx, count):
    if idx == len(arr):
        print(' '.join([str(x) for x in arr]))
        count[0] += 1
        return 

    i = idx
    while i < len(arr):
        arr[idx], arr[i] = arr[i], arr[idx]
        permutation(arr, idx + 1, count)
        arr[idx], arr[i] = arr[i], arr[idx]
        i += 1
    
if __name__ == "__main__":
    main()

File: Week01/d3_permutation.c
/**
 * Author: Penut Chen
 * Date: 2019-6-29
 * Description: 
 *   Practice of permutation function.
 */

#include <stdio.h>
#include <stdlib.h>

int* generate_array(int);
void swap(int*, int*);
void permutation(int*, int, int, int*);

int main() {
    int size = 4, count = 0;
    int *arr = generate_array(size);
    
    permutation(arr, size, 0, &count);
    printf("Total %d\n", count);
    free(arr);
    
    return 0;
}

int* generate_array(int size) {
    int *arr = (int*)malloc(sizeof(int) * size);
    for (int i = 0; i < size; i++)
        arr[i] = i + 1;
    return arr;
}

void permutation(int *arr, int size, int idx, int *count) {
    if (idx == size) {
        for (int i = 0; i < size; i++)
            printf("%d ", arr[i]);
        printf("\n");
        (*count) += 1;
        return ;
    }
    
    for (int i = idx; i < size; i++) {
        swap(&arr[i], &arr[idx]);
        permutation(arr, size, idx + 1, count);
        swap(&arr[i], &arr[idx]);
    }
}

void swap(int *a, int *b) {
    int t = *a;
    *a = *b;
    *b = t;
}